### PR TITLE
Allowing outputing SVG in different streams (terminal/file/string)

### PR DESCRIPTION
**Pull request by yixuan**
This PR is an attempt for the proposal in https://github.com/hadley/svglite/issues/40.

Apart from saving SVG to a file in hard disk, this patch adds two additional way to save the result: to print the content of SVG in the R terminal, and to save the SVG as a string.

The design consists of the following aspects:
- The C-style printing functions are replaced by C++ style streams.
- In this way, we can use a common parent class `std::ostream` to unify different ways of output. Terminal stream is handled by `Rcpp::Rcout`, file stream by `std::ofstream` and string stream by `std::ostringstream`.
- To tell the device that SVG should be printed to the R terminal, use `svglite(file = ":terminal:")` in the driver function; similarly, specify `svglite(file = ":string:")` for writing to string stream.
- The string stream version is a bit tricky, since we cannot expect `dev.off()` to return the wanted string. Instead, I added another function `stringSVG()` to extract it. The usage will look like
```r
svglite(":string:")
plot.new()
dev.off()
svgstr <- stringSVG()
cat(svgstr)
```
`stringSVG()` always returns the last plot created by `svglite(":string:")`.

Any further comments or suggestions?

### CODE REVIEW

**Comment by jeroenooms** [comment]
Very nice.

**Comment by jeroenooms** [comment]
Is there a way to make `stringSVG()` return intermediate svg before calling `dev.off()` so that we can support live preview while generating a plot?

**Comment by timelyportfolio** [comment]
:+1: I would very much like to see this implemented.  Thanks @yixuan for all your work on this.

**Comment by timelyportfolio** [comment]
If this is accepted, should we alter `inlineSVG` to use the `stringSVG()` method to avoid tempfile/delete tempfile dance?

**Comment by yixuan** [comment]
@jeroenooms I think it's possible, but I need to take some investigation. I'll update the status here.
@timelyportfolio Thanks!

Wish all of you a happy new year!

**Comment by noamross** [comment]
I had missed this and realized I implemented some what @jeroenooms in #53.  The goal there was to support live preview, too.

@yixuan, I did this by writing `</svg>`, flushing the graphics stream, and then setting the location back 7 bytes after each callback.  This way the document is alwyas valid SVG and the next command will overwrite the `</svg>`. 

```
fputs("</svg>\n", svgd->file)
fflush(svgd->file);
fseek(svgd->file, -7, SEEK_CUR);
```

There are equivalents of each of these for `ofstream`.

File: src/devSVG.cpp [hunk 1]
```diff
@@ -80,79 +118,90 @@ inline std::string fontname(const char* family_, int face) {
...
 
-inline void write_escaped(FILE* f, const char* text) {
+inline void write_escaped(std::ostream* stream, const char* text) {
   for(const char* cur = text; *cur != '\0'; ++cur) {
     switch(*cur) {
-    case '&': fputs("&amp;", f); break;
-    case '<': fputs("&lt;",  f); break;
-    case '>': fputs("&gt;",  f); break;
-    default:  fputc(*cur,    f);
+    case '&': (*stream) << "&amp;"; break;
```
Comment (hadley) [inline src/devSVG.cpp]: It feels like there must be a better way to do this, but the write method doesn't take a std::string. @jjallaire, do you have any suggestions?

File: src/devSVG.cpp [hunk 2]
```diff
@@ -241,55 +290,59 @@ void svg_clip(double x0, double x1, double y0, double y1, pDevDesc dd) {
 }
 
 void svg_new_page(const pGEcontext gc, pDevDesc dd) {
+BEGIN_RCPP
```
Comment (hadley) [inline src/devSVG.cpp]: Why is this necessary?

File: src/devSVG.cpp [hunk 3]
```diff
@@ -36,16 +49,41 @@ class SVGDesc {
...
-    file = fopen(R_ExpandFileName(filename.c_str()), "w");
-  }
 
-  bool ok() const {
-    return file != NULL;
+    if (filename == ":terminal:") {
+      stream = (std::ostream*) &(Rcpp::Rcout);
+      stream_type = 0;
+    } else if (filename == ":string:") {
+      stream = new std::ostringstream();
```
Comment (hadley) [inline src/devSVG.cpp]: I think we'd be better off avoiding explicit memory management, and instead using a `boost::shared_ptr()`.

File: src/devSVG.cpp (hunk 2 above)
Comment (yixuan) [inline src/devSVG.cpp]: Oh, this is because `Rcpp::stop()` throws an exception, and it must be captured by Rcpp. `svg_new_page()` is a device function, which will be directly called by R and won't be protected by Rcpp without such macros.

File: src/devSVG.cpp (hunk 3 above)
Comment (yixuan) [inline src/devSVG.cpp]: Good point.

File: src/devSVG.cpp (hunk 2 above)
Comment (hadley) [inline src/devSVG.cpp]: In that case I'd rather stick to calling `Rf_error`

Comment (yixuan) [inline src/devSVG.cpp]: I see. I did that since I saw there was a discussion by @kevinushey: http://lists.r-forge.r-project.org/pipermail/rcpp-devel/2015-March/008501.html

> Calling `Rf_error()` directly from C++
can result in undefined behaviour -- this is because calling
`Rf_error()` will result in a `longjmp`, which bypasses the
destructors of any objects on the stack (thereby potentially leaking
memory, but in general, causing undefined behaviour).

Comment (jjallaire) [inline src/devSVG.cpp]: Yeah if we call `Rf_error` directly we will indeed skip C++ destructors. This approach allows us to catch the C++ exception and then convert it to an `Rf_error` for the R C layers benefit.

Comment (hadley) [inline src/devSVG.cpp]: Oh yeah, good point.  This code probably would've have been better as pure C, but we might as go fully towards C++, since there are no performance implications here.

File: src/devSVG.cpp (hunk 3 above)
Comment (hadley) [inline src/devSVG.cpp]: If you implement that, I think the rest of the PR is good to merge

Comment (yixuan) [inline src/devSVG.cpp]: Yeah. I'll think about it and see if we have an elegant solution.

Comment (hadley) [inline src/devSVG.cpp]: I'd like to get this to CRAN by the end of next week, so if you can't think of anything elegant, we can stick with this and figure it out later.

Comment (yixuan) [inline src/devSVG.cpp]: Sure. I'll get you back shortly, hopefully by this weekend.

**Comment by yixuan** [comment]
The current implementation avoids dynamic memory allocation by defining member variables `strstream` and `filestream`, which are intended to handle string and file respectively, but only one of them will be used, depending on the stream type detected from the filename.

These variables will be handled by the destructor automatically, so we don't need to `new` and `delete` by ourselves.

**Comment by hadley** [comment]
Thanks!

**Comment by hadley** [comment]
Hmmm, `:terminal:` doesn't interact well with always producing valid SVG. It also doesn't seem that useful to me - why did you add it?

**Comment by yixuan** [comment]
I see. I thought it would be useful for `knitr` to produce in-document SVG graphics, using the "as is" mode.

**Comment by hadley** [comment]
Hmmmmm, but you can already use svglite with knitr using the regular plot options.
